### ISSUE TITLE
Transition to new server and library code

### ISSUE THREAD

**Issue by codefromthecrypt** [openzipkin/zipkin#1119]
After settling on a single codebase in #1071, and the release of zipkin-java 1.0, we have some consolidation steps to take. These include at least the following:
- Determine how to migrate the zipkin-java code into this repository, whether or not we preserve history from zipkin-java.
- Decide what to do with packages (ex. zipkin-java is under the group id io.zipkin.java at the moment)
- break out issues to consolidate docker and decomission other repos.

An alternative discussed was to rename this repo and archive it. While simpler mechanics, this was initially thrown out as this repo has a lot of followers. Moreover, it is unclear if you can rename a repo to to archive-X, then have a different repo become X.

**Comment by codefromthecrypt** [comment]
so here's an idea. Right now, the only code that's needed and not in zipkin-java is zipkin-ui. 

## Preparation: move zipkin-ui to zipkin-java and stop development here

Regardless, first step we should take is move zipkin-ui to zipkin-java and stop affecting master here:
1. Merge pull requests and cut a final release of zipkin-ui here
2. Migrate zipkin-ui to zipkin-java repo (via `git filter-branch` and friends to preserve history)
   - implies converting the build script to the maven alternative
3. cut zipkin-ui from the zipkin-java repo

At this point, everything is in sync and we have a couple options to proceed.

## Option 1: shuffle repository names

The simplest option would be to rename zipkin to something else (zipkin-archived or zipkin-legacy) and then rename zipkin-java to zipkin.

### Pros
- no chance to screw up forks (they basically can continue off master and/or be abandoned)
- irrelevant issues and old pull requests aren't imported to the new codebase
- current issues in zipkin-java are retained

### Cons
- Not sure if the rename shuffle will work or not. Ex can you rename a github repository to a historical name of another?
- Some issues like UI ones, or the new model will need to be recreated
- zipkin-java has less stars and watchers.
  - These would need to be organically regrown
- zipkin-java has less collaborators.
  - Folks who've done historical work in zipkin won't show up in the commit log until they do more work.

## Option 2: git-fu

Another approach would be to rename master to scala, push master from the java codebase, then turn off zipkin-java

### Pros
- zipkin stars, watchers, and collaborator stats remain in place
- issues that don't relate to the scala implementation don't need to be recreated.

### Cons
- folks with existing forks off master on either side will be disrupted
  - The act of turning off zipkin-java abandons relevant forks off the current codebase
  - The act of renaming master here disrupts any forks off the old codebase
- closed pull requests and issues have little to do with the current codebase or build.
  - Basically search is irrelevant, as most issues will refer to things that never affected the java port, ex finagle, scrooge, anorm, sbt, gradle...

---

#### Adrian's vote: Option 1: shuffle

The more I think about it, the less I like option 2. I feel it disrupts both sides with technical gotchas. While I do like the idea of memorializing past contributions and stars, I hope we've made a project that people will star again, and contribute to again. I'll certainly continue to contribute effort to the socializing of zipkin such that stars might be reborn faster. I'd also offer effort to recreate issues that are still relevant.

**Comment by codefromthecrypt** [comment]
cc'ing people who've contributed code in the last month to either repo:
- zipkin(-scala): @NegatioN @esbie @yurishkuro 
- zipkin-java: @sveinnfannar @dsyer @abesto @eirslett @michaelsembwever 

(note zipkin-ui history will be preserved as a part of the migration regardless)

**Comment by yurishkuro** [comment]
what about moving scala code to a branch in `zipkin` repo?  If people have forks which are reasonably up to date, can they just rebase from the `scala` branch?

**Comment by codefromthecrypt** [comment]
How would moving the scala code into a branch help with the problems of
lost stars etc? If anything, it would seem to confuse the fact that it is
not a supported codebase anymore.

**Comment by yurishkuro** [comment]
I meant as a flavor of option 2, i.e. consolidate both java and scala impls in zipkin repo, in master and scala branches respectively. It may address the first con (with rebase), but not the second. 

The second con is a bit questionable, as there may be many things in closed issues that shed light as to why certain things are done one way or another, especially if java impl just copied the original logic. But I hardly ever used search on the old issues/PRs, so I'm not the right person to judge that.

**Comment by codefromthecrypt** [comment]
ah ok. Makes more sense now.

**Comment by codefromthecrypt** [comment]
> The second con is a bit questionable, as there may be many things in
> closed issues that shed light as to why certain things are done one way or
> another, especially if java impl just copied the original logic. But I
> hardly ever used search on the old issues/PRs, so I'm not the right person
> to judge that.
> 
> I might be a decent person to judge having participated in most issues in
> both repos in the last year (as well rewriting the code).

There's well over 1000 issues in the old repo, and quite many pertain to
specifics of finagle, scrooge, sbt, gradle, etc. The code and code docs in the
zipkin-java repo is far better documented than the old ones, and the pull
request and commit descriptions etc are far more elaborate, too. While
there are some good issues in the old repo, the majority are not going to
ease understanding of the new codebase. If given a choice, I'd stick with
the ones that directly caused the current codebase.

**Comment by eirslett** [comment]
I'm voting for yurishkuro's solution! The variant of 2.

**Comment by codefromthecrypt** [comment]
I'm OK with either, especially as either path is towards the end of tasks such as this :p

So to be clear, the git-fu would be something like this..

``` bash
# after git tag 1.40.2 is pushed and released.. we have a cease-fire on the repo

# First, from zipkin repo, rename master to scala
$ git branch -m master scala
$ git push origin scala

# Now, from zipkin-java repo, push master to the zipkin repo
$ git remote add zipkin https://github.com/openzipkin/zipkin.git
$ git push -f zipkin master

# Next, from the zipkin repo, fix travis credentials and cut a patch release to test it
```

**Comment by codefromthecrypt** [comment]
@gregwhitaker @anuraaga fyi, if we take option 2, you'd need to rebase your forks off a different repo (even though master will end up with the same code as zipkin-java)

**Comment by codefromthecrypt** [comment]
mildly revised the option 2 desc to include the master rename

**Comment by codefromthecrypt** [comment]
Ps something I didn't mention earlier is a pro in-favor of option 2 is that
the UI issues are retained. Since the code is the same, this is helpful.

Also, I thought of a way to balance the problem of losing zipkin-java's
issues. There are probably less than 10 with lots of comments. Maybe we can
harvest the comments into a doc or something.

**Comment by eirslett** [comment]
We can delete all the code from the zipkin-java and leave only a readme. Then we keep the issues. And we leave it in that state for a couple of months. We make a single "first commit" commit and force push it to the master branch, which will wipe the git history, so that it's not possible to merge or pull/push to it inadvertently.

**Comment by eirslett** [comment]
There are also some open source scripts written to automate github issue moving. It won't replace author/timestamps/metadata, but put that in the issue texts.

**Comment by codefromthecrypt** [comment]
Thanks. Will add it to the plan. If I don't hear strongly otherwise, we'll
go with option 2.

**Comment by codefromthecrypt** [comment]
here's some related work in docker land, basically renaming the java image to "zipkin" as it doesn't need qualification since it includes collector and query functionality https://github.com/openzipkin/docker-zipkin-java/pull/9

**Comment by codefromthecrypt** [comment]
I'm starting this stuff now

**Comment by codefromthecrypt** [comment]
Executed the following:

``` bash
# First, from zipkin repo, rename master to scala
$ git branch -m master scala
$ git push origin scala

# Now, from zipkin-java repo, push master to the zipkin repo
$ git remote add zipkin https://github.com/openzipkin/zipkin.git
$ git push -f zipkin master
```

**Comment by codefromthecrypt** [comment]
verified master looks alright and scala branch is in-tact

https://github.com/openzipkin/zipkin
https://github.com/openzipkin/zipkin/tree/scala

**Comment by codefromthecrypt** [comment]
renamed the old bintray repository to zipkin-archived so that we don't accidentally overwrite versions

**Comment by codefromthecrypt** [comment]
renamed the bintray repository zipkin-java to zipkin

**Comment by codefromthecrypt** [comment]
updated the travis things and links https://github.com/openzipkin/zipkin/commit/2b50556dbb940d39fb6b4ae6dfc04a1c042ad143

**Comment by codefromthecrypt** [comment]
moved release tags for 1.0.0 and 1.1.0 and backfilled release notes

**Comment by codefromthecrypt** [comment]
@gregwhitaker @anuraaga note! what was zipkin-java is now here. please update your refs accordingly

**Comment by codefromthecrypt** [comment]
created a branch on zipkin-java called cutover and force pushed a README that redirects here

**Comment by codefromthecrypt** [comment]
done!
